File: src/science/amberfall/dumbo/Dumbo.java
package science.amberfall.dumbo;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.URL;
import java.util.List;
import java.util.Random;

import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.command.Command;
import org.bukkit.command.CommandSender;
import org.bukkit.command.ConsoleCommandSender;
import org.bukkit.configuration.file.YamlConfiguration;
import org.bukkit.entity.Player;
import org.bukkit.event.player.AsyncPlayerChatEvent;
import org.bukkit.plugin.Plugin;
import org.bukkit.plugin.java.JavaPlugin;
import org.bukkit.scheduler.BukkitRunnable;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;

public class Dumbo extends JavaPlugin implements Listener {

    static Dumbo plugin;

    public static Plugin getPlugin() {
        JavaPlugin plugin = Dumbo.plugin;
        return plugin;
    }

    @Override
    public void onEnable() {

        plugin = this;

        if (Bukkit.getServer().getVersion().contains("git-Spigot")) {

            ConsoleCommandSender console = Bukkit.getServer().getConsoleSender();

            console.sendMessage("[Dumbo]" + ChatColor.RED + "We highly recommend that you use Paper over Spigot, as they have been hostile to me personally and the development of this plugin. Paper offers many performance improvements and optimizations over Spigot. Please download and install Paper from https://paperci.emc.gs to use this plugin.");
            Bukkit.getPluginManager().disablePlugin(this);

        } else {
            createConfig();
            Bukkit.getServer().getPluginManager().registerEvents(this, this);
        }

    }

    @Override
    public void onDisable() {}

    private void createConfig() {
        try {

            if (!getDataFolder().exists()) {
                getDataFolder().mkdirs();
            }

            File file = new File(getDataFolder(), "config.yml");

            if (!file.exists()) {
                getLogger().info("Fetching quotes from GitHub..");
                Bukkit.getScheduler().runTaskAsynchronously(Dumbo.getPlugin(), new Runnable() {

                    @Override
                    public void run() {

                        InputStream in = null;

                        try { in = new URL("https://raw.githubusercontent.com/sweepyoface/dumbo/master/quotes.yml").openStream();
                        } catch (IOException e) {
                            e.printStackTrace();
                        }

                        YamlConfiguration yaml = YamlConfiguration.loadConfiguration(new InputStreamReader( in ));
                        Bukkit.getScheduler().runTask(Dumbo.getPlugin(), new Runnable() {

                            @Override
                            public void run() {
                                String color = "&6";
                                Dumbo.getPlugin().getConfig().addDefault("color", color);
                                Dumbo.getPlugin().getConfig().addDefault("quotes", yaml.getStringList("quotes"));
                                Dumbo.getPlugin().getConfig().options().copyDefaults(true);
                                saveConfig();
                                reloadConfig();
                                getLogger().info("Done!");
                            }

                        });
                    }
                });

            }

        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public String randomQuote() {
        Random random = new Random();
        List <String> quotes = this.getConfig().getStringList("quotes");
        if (quotes.size() == 0) {
            return "false";
        } else {
            Integer quote = random.nextInt(quotes.size());
            return quotes.get(quote);
        }
    }

    @Override
    public boolean onCommand(CommandSender sender, Command cmd, String label, String[] args) {
        Player player = (Player) sender;
        if (label.equalsIgnoreCase(("dumbo")) && sender instanceof Player) {
            if (args.length == 0) {
                if (player.hasPermission("dumbo.quote") || player.isOp()) {
                    String quote = randomQuote();
                    if (quote == "false") {
                        player.sendMessage(ChatColor.DARK_RED + "Error: Quotes are still being initialized, please wait or contact a server administrator.");
                    } else {
                        Bukkit.getServer().broadcastMessage(ChatColor.translateAlternateColorCodes('&', this.getConfig().getString("color")) + quote);
                    }
                }
            } else if (args[0].equalsIgnoreCase("reload")) {
                if (player.hasPermission("dumbo.reload") || player.isOp()) {
                    this.reloadConfig();
                    player.sendMessage(ChatColor.GREEN + "Dumbo config reloaded!");
                } else {
                    player.sendMessage(ChatColor.DARK_RED + "You do not have access to that command.");
                }
            } else {
                player.sendMessage(ChatColor.RED + "Unknown argument. See /help dumbo");
            }
        }
        return true;
    }

    @EventHandler
    public void onChat(AsyncPlayerChatEvent ev) {
        if (!ev.isCancelled()) {
            String[] msgArray = ev.getMessage().trim().split("\\s+");
            if (msgArray[0].equalsIgnoreCase(".dumbo")) {
                Player player = (Player) ev.getPlayer();
                if (player.hasPermission("dumbo.quote") || player.isOp()) {
                    String quote = randomQuote();
                    if (quote == "false") {
                        player.sendMessage(ChatColor.DARK_RED + "Error: Quotes are still being initialized, please wait or contact a server administrator.");
                    } else {
                        new BukkitRunnable() {
                            @Override
                            public void run() {
                                Bukkit.getServer().broadcastMessage(ChatColor.translateAlternateColorCodes('&', Dumbo.getPlugin().getConfig().getString("color")) + quote);
                            }
                        }.runTaskLater(this, 1);
                    }
                } else {
                    player.sendMessage(ChatColor.DARK_RED + "You do not have access to that command.");
                }
            }
        }
    }
}

File: src/main/kotlin/science/amberfall/dumbo/Dumbo.kt
package science.amberfall.dumbo

import org.bukkit.ChatColor
import org.bukkit.command.Command
import org.bukkit.command.CommandSender
import org.bukkit.configuration.file.YamlConfiguration
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.player.AsyncPlayerChatEvent
import org.bukkit.plugin.java.JavaPlugin
import java.io.File
import java.io.InputStreamReader
import java.net.URL
import java.util.*

class Dumbo : JavaPlugin(), Listener {

    override fun onEnable() {
        createConfig()
        server.pluginManager.registerEvents(this, this)

        if(server.version.contains("git-Spigot")) {
            logger.info("We highly recommend that you use Paper over Spigot, as they have been hostile to me personally and the development of this plugin. Paper offers many performance improvements and optimizations over Spigot. Please download and install Paper from https://paperci.emc.gs to use this plugin.".red())
            server.pluginManager.disablePlugin(this)
        }
    }

    private fun createConfig() {
        if(!dataFolder.exists()) {
            dataFolder.mkdirs()
        }
        val file = File(dataFolder, "config.yml")
        if(!file.exists()) {
            logger.info("Fetching quotes from GitHub...")

            val stream = URL("https://raw.githubusercontent.com/sweepyoface/dumbo/master/quotes.yml")
                    .openStream()
            val yaml = YamlConfiguration.loadConfiguration(InputStreamReader(stream))

            val colour = "&6"

            config.addDefault("color", colour)
            config.addDefault("quotes", yaml.getStringList("quotes"))

            config.options().copyDefaults(true)
            saveConfig()
            reloadConfig()

            logger.info("Done!")
        }
    }

    fun randomQuote(): String {
        val rand = SplittableRandom()
        val quotes = config.getStringList("quotes")
        val quote = rand.nextInt(quotes.size)

        return quotes[quote]
    }

    override fun onCommand(sender: CommandSender, cmd: Command, label: String, args: Array<String>): Boolean {
        if(sender is Player) {
            if(label.equals("dumbo", true)) {
                if(args.isEmpty()) {
                    if(sender.hasPermission("dumbo.quote") || sender.isOp) {
                        server.broadcastMessage(
                                ChatColor.translateAlternateColorCodes('&', config.getString("color") + randomQuote()))

                    }else if(args[0].equals("reload", true)) {
                        if(sender.hasPermission("dumbo.reload") || sender.isOp) {
                            this.reloadConfig()
                            sender.sendMessage("Dumbo config reloaded".green())
                        }
                    }else {
                        sender.sendMessage("Unknown argument. See /help dumbo".red())
                    }
                }
            }
        }
        return true
    }

    @EventHandler
    fun onChat(ev: AsyncPlayerChatEvent) {
        if(!ev.isCancelled) {
            val msgArr = ev.message.trim().split("\\s+".toRegex())
            if(msgArr[0].equals(".dumbo", true)) {
                val p = ev.player
                if(p.hasPermission("dumbo.quote") || p.isOp) {
                    server.broadcastMessage(
                            ChatColor.translateAlternateColorCodes('&', config.getString("color") + randomQuote())
                    )
                }else {
                    p.sendMessage("You do not have access to that command.".darkRed())
                }
            }
        }
    }
}

fun String.red(): String = "${ChatColor.RED}$this"
fun String.green(): String = "${ChatColor.GREEN}$this"
fun String.darkRed(): String = "${ChatColor.DARK_RED}$this"
